common/determined_common/_logging.py

```diff
@@ -1,16 +1,207 @@
+import contextlib
+import json
 import logging
 import sys
+import threading
+import time
+from typing import Any, Callable, Dict, Generator, Optional, Union
 
 
-def set_logger(debug_enabled: bool) -> None:
-    root = logging.getLogger()
-    root.setLevel(logging.DEBUG if debug_enabled else logging.INFO)
+class _Loggers:
+    def __init__(self) -> None:
+        self._storage = logging.getLogger("det.storage")
+        self._harness = logging.getLogger("det.harness")
+        self._resources = logging.getLogger("det.resources")
+        self._all_loggers = (self._storage, self._harness, self._resources)
 
-    for hdlr in root.handlers:
-        root.removeHandler(hdlr)
+    @property
+    def storage(self) -> logging.Logger:
+        return self._storage
 
-    handler = logging.StreamHandler(sys.stdout)
-    handler.setLevel(logging.DEBUG if debug_enabled else logging.INFO)
-    formatter = logging.Formatter("%(levelname)s: %(message)s")
-    handler.setFormatter(formatter)
-    root.addHandler(handler)
+    @property
+    def harness(self) -> logging.Logger:
+        return self._harness
+
+    @property
+    def resources(self) -> logging.Logger:
+        return self._resources
+
+
+# Loggers are normally global in python, and it would be a bit unexpected to behave differently.
+log = _Loggers()
+
+
+class _ProfilingThread(threading.Thread):
+    """
+    Call a profiling function periodically on a separate thread.
+    """
+
+    def __init__(self, period: float, profile_fn: Callable) -> None:
+        self.period = period
+        self.profile_fn = profile_fn
+        self.quitting = False
+        self.cond = threading.Condition()
+        super().__init__()
+
+    def run(self) -> None:
+        with self.cond:
+            while True:
+                self.cond.wait(timeout=self.period)
+                if self.quitting:
+                    break
+                self.profile_fn()
+
+    def quit(self) -> None:
+        self.quitting = True
+        with self.cond:
+            self.cond.notify_all()
+
+
+def read_level(level: Optional[str], default: int, limit: int = 0) -> int:
+    return max(logging._checkLevel(level if level is not None else default), limit)  # type: ignore
+
+
+class DebugConfig:
+    def __init__(
+        self,
+        root_log_level: Optional[str] = None,
+        storage_log_level: Optional[str] = None,
+        harness_log_level: Optional[str] = None,
+        debug_all_workers: bool = False,
+        horovod_verbose: bool = False,
+        nccl_debug: Optional[str] = None,
```

Comment:
I cannot remember any time that nccl logs were useful to me.  Have you seen them be useful often?

Reply:
To be clear, I want `debug: true` to be useful to let users self-service themselves easily.  I consider the NCCL debug info to be noisy and rarely useful to the average user, that's why I didn't include it.  That's why I also chose not to dump the stack trace every 30 seconds... that's really only useful for detecting why processes hang (which is much less common than when we added it) and I've seen users who are very confused to see stack traces when they set `debug: true`.  Confused users means they have to come to use which is against the goal of letting average users self-service.

Reply:
Oh actually, you were talking about the default behavior, and I was talking about the `debug: true` behavior, but I feel about the same in both cases... I find the NCCL info highly un-useful.

Reply:
I have actually found NCCL INFO useful on multiple occasions before debugging networking issues, including that GPU interconnects are being used correctly as well as providing overall proof that NCCL is being used at all. We should definitely have it on for `debug`: `true`, but I would also be strongly in favor in including it by default as well, since after startup they do not output logs.

Reply:
NCCL INFO is also hugely useful when debugging hangs and network interface problems. I definitely think they should be present in `debug: true`. I don't have an opinion on whether they should be included by default.

Reply:
We can definitely turn it on for `debug: true`.  @aaron I would still like to default to the `nccl_debug: null` case because:
1. I think that on principle you should be allowed to turn off debug log output that you don't want
2. I am not sure of a clean way to allow the `nccl_debug: null` case while still defaulting to "INFO" without having doing full experiment config expansion in the master (at least, not without the full experiment config project), which I dearly would like to avoid.
3. If we did allow `nccl_debug: null` at all it would seem very strange for that not to be the default.  You've convinced me it is useful for debugging often enough that I'm happy to have it with `debug: true`, but it seems like after users initially get their networking in order most users won't have to look at the NCCL logs again.  And unnecessary logs is a bad user experience.

Reply:
Yeah this is another reason I'm not a huge fan of setting defaults in the harness vs. master the way we do for everything else. Another approach we could take is to require users to set it to `disable` in order to turn the logs off. 

Turning it off by default doesn't seem like the end of the world. What I am worried about even more than debugging is the questions of "is anything happening with my training or is it stuck", the NCCL logs appearing indicate that your training is indeed occurring while with them off users are waiting for a step to complete, which sometimes takes a while.